### PR TITLE
[AIRFLOW-2330] Do not append destination prefix if not given

### PR DESCRIPTION
**Pull request by berislavlopac**
Addresses [AIRFLOW-2330](https://issues.apache.org/jira/browse/AIRFLOW-2330).

### Description
The change introduces a check whether the `destination_object` parameter was given when using a wildcard (`*`) in the `source_object` parameter. If `destination_object` resolves to `False`, e.g. when not given (the default value is `None`) or is an empty string, the `destination_object` path is simply the `source_object` path.

It also includes some minor updates to docs and log messages.

### CODE REVIEW

**Comment by Fokko** [comment]
This will enable you to copy something from the source, to the source? Could you please elaborate, I find the logic hard to follow.

**Comment by berislavlopac** [comment]
When using the wildcard, you're implicitly copying multiple files (e.g. `foo/*` means all files that start with `foo/`). The `destination_object` doesn't make sense in that context, so it was used to specify an extra prefix (often called a "folder" or "subfolder", although GCS, like S3, doesn't have an inherent hierarchical organisation) to be prepended to the destination paths.

Here's an example: let's assume that we have two files in a bucket called `some_source_bucket`: `foo/file1.txt` and `foo/file2.txt`. Now, say that we do:

```
copy_foo_files = GoogleCloudStorageToGoogleCloudStorageOperator(
    dag=dag,
    task_id='copy_foo_files',
    source_bucket='some_source_bucket',
    destination_bucket='some_destination_bucket',
    source_object='foo/*',
    destination_object='bar/baz'
)
```

This will copy files:

```
gs://some_source_bucket/foo/file1.txt
gs://some_source_bucket/foo/file2.txt
```
into:
```
gs://some_destination_bucket/bar/baz/foo/file1.txt
gs://some_destination_bucket/bar/baz/foo/file2.txt
```
Hope this makes sense. :smile: 

The issue being fixed here was that, if you omitted `destination_object` or sent an empty string, the files would be copied into:
```
gs://some_destination_bucket/None/foo/file1.txt
gs://some_destination_bucket/None/foo/file2.txt
```
or
```
gs://some_destination_bucket//foo/file1.txt
gs://some_destination_bucket//foo/file2.txt
```
Since the GCS is not a hierarchical filesystem, the latter is not the same as
```
gs://some_destination_bucket/foo/file1.txt
gs://some_destination_bucket/foo/file2.txt
```

**Comment by berislavlopac** [comment]
Ideally, we could have a system that would allow building complex source and destination paths for multiple files, but that's a bigger fish to fry, and outside the scope of this ticket anyway.

**Comment by Fokko** [comment]
Thanks @berislavlopac for the examples. It makes sense now to me. Maybe a good idea to turn one or two of these examples into an unit-test. This is some delicate code since we're possibly moving data here.

**Comment by berislavlopac** [comment]
Tests added.

**Review by Fokko — APPROVED** [review]
Thanks @berislavlopac for adding tests. Merging!
